components/higher-order/with-dragging/index.js

```diff
@@ -0,0 +1,163 @@
+		 * @param  {Object} event     The non-custom DragEvent.
+		 */
+		onDragEnd( event ) {
+			const { elementId, cloneNodeId } = this.state;
+			const element = document.getElementById( elementId );
+			const cloneWrapper = document.getElementById( cloneNodeId );
+
+			if ( element && cloneWrapper ) {
+				// Remove clone.
+				element.parentNode.removeChild( cloneWrapper );
+			}
+
+			// Reset cursor.
+			document.body.classList.remove( 'dragging' );
+
+			this.setState( { elementId: null, cloneNodeId: null } );
+
+			document.removeEventListener( 'dragover', this.onDragOver );
+			event.stopPropagation();
+		}
+
+		/*
+		 * Updates positioning of element clone based on mouse movement during dragging.
+		 * @param  {Object} event     The non-custom DragEvent.
+		 */
+		onDragOver( event ) {
+			const { cloneNodeId } = this.state;
+			const cloneWrapper = document.getElementById( cloneNodeId );
+
+			cloneWrapper.style.top =
+				`${ parseInt( cloneWrapper.style.top, 10 ) + parseInt( event.clientY, 10 ) - parseInt( this.cursorTop, 10 ) }px`;
+			cloneWrapper.style.left =
+				`${ parseInt( cloneWrapper.style.left, 10 ) + parseInt( event.clientX, 10 ) - parseInt( this.cursorLeft, 10 ) }px`;
+
+			// Update cursor coordinates.
+			this.cursorLeft = event.clientX;
+			this.cursorTop = event.clientY;
+		}
+
+		/**
+		 *  - Clones the current element and spawns clone over original element.
+		 *  - Adds a fake temporary drag image to avoid browser defaults.
+		 *  - Sets transfer data.
+		 *  - Adds dragover listener.
+		 * @param  {Object} event     The non-custom DragEvent.
+		 * @param  {string} uid       The unique identifier of the element to be dragged.
+		 * @param  {string} elementId The HTML id of the element to be dragged.
+		 * @param  {number} order     The current index of the element to be dragged.
+		 * @param  {string} type      A unique constant identifying the type of dragging operation.
+		 */
+		onDragStart( event, uid, elementId, order, type ) {
+			const element = document.getElementById( elementId );
+			const elementWrapper = element.parentNode;
+			const elementRect = element.getBoundingClientRect();
+			const elementTopOffset = parseInt( elementRect.top, 10 );
+			const elementLeftOffset = parseInt( elementRect.left, 10 );
+			const clone = element.cloneNode( true );
+			const cloneWrapper = document.createElement( 'div' );
+			const cloneNodeId = `clone-wrapper-${ element.id }`;
+			const dragImage = document.createElement( 'div' );
+
+			this.setState( { cloneNodeId, elementId } );
+
+			// Set a fake drag image to avoid browser defaults. Remove from DOM right after.
+			if ( 'function' === typeof event.dataTransfer.setDragImage ) {
+				dragImage.id = `drag-image-${ element.id }`;
+				dragImage.classList.add( this.config.dragImageClass );
+				document.body.appendChild( dragImage );
+				event.dataTransfer.setDragImage( dragImage, 0, 0 );
+				setTimeout( ( ( _dragImage ) => () => {
```

Comment:
Can you explain why we need a timeout? Also, the second argument is useless I guess since it's the default. Also, any particular reason you're using an IIFE instead of a simple callback? These could be harder to understand.

Reply:
Further, if we really do need a timeout, we should use `withSafeTimeout`, since this one is "unsafe" in that it doesn't cancel if the component unmounts. Based on the current logic, it is probably not error-prone, but causes concern for future maintainability where it might start assuming to have access to component properties.

https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-safe-timeout

Reply:
This is definitely needed here afaik. I've had a bit of trouble with this particular logic sequence here. It was a much larger issue initially as I was using the `setDragImage` call to also handle the block preview that you see when dragging. Originally, I would clone the block, set it as a drag image, and then remove the clone. Then the drag image would be what you see in the preview. The problem was IE does not support the `setDragImage` call at all, hence I ended up cloning the block and moving it around as the mouse moves (via `dragOver`), and abandoning the `setDragImage` call alltogether. That is the only way I could possibly have the design consistent with IE.

Anyway, although it's not easy to see the issue here (cause there's nothing visible) - the problem that I had with the sequence `clone element -> attach to DOM -> call setDragImage -> remove clone` is it would always fail on `setDragImage`. The clone was removed from the DOM before the drag image was set. Executing the last part in a timeout was the only I found to get this working. I have tried to investigate `setDragImage` to see if there is any concept of a timeout in there, but I had no luck.